### PR TITLE
Offline VAD+ASR tutorial

### PR DESCRIPTION
**Pull request by fayejf**
# What does this PR do ?

This PR adds functions and a tutorial about how to perform offline VAD before ASR.
Pipeline: audio -> [VAD] -> speech segments -> [ASR] -> segmented ASR output -> stitched ASR output

**Collection**:  ASR

# Changelog 
- Transcribe partial audio (for example speech segments) with manifest.
- Stitch segmented ASR output.
- Construct the manifest that aligns stitched ASR outputs and ground truth text. 
- And a walkthrough tutorial for the whole process. 

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/asr/transcribe_speech.py [hunk 1]
```diff
@@ -151,10 +151,18 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
     else:
         # get filenames from manifest
         filepaths = []
+        if os.stat(cfg.dataset_manifest).st_size == 0:
+            logging.warning(f"The input dataset_manifest {cfg.dataset_manifest} is empty. Exiting!")
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:155]: Logging.error is better here. Also, cool check to test for empty manifest.

**Review by titu1994 — CHANGES_REQUESTED** [review]
This is not a good change. the .transcribe() method will only work on python string paths. It is not meant for manifest handling and setting up of lengths etc. That's the point of the transcribe_speech.py script - to provide a higher level API over the transcribe method (supporting manifest etc). 

Please revert these changes to transcribe - especially the requirement of num files etc. Such things are not meant for transcribe. They are meant for higher level script to work on transcribe.

File: examples/asr/transcribe_speech.py [hunk 2]
```diff
@@ -93,6 +94,86 @@ class TranscriptionConfig:
     rnnt_decoding: RNNTDecodingConfig = RNNTDecodingConfig(fused_batch_size=-1)
 
 
+def transcribe_partial_audio(
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:97]: Where's the equivalent for RNNT ? If it's not supported, raise an error please. You can do this by asserting that obj is instance of only CTC classes.

File: examples/asr/transcribe_speech.py [hunk 3]
```diff
@@ -151,10 +232,22 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
         filepaths = []
+        if os.stat(cfg.dataset_manifest).st_size == 0:
+            logging.error(f"The input dataset_manifest {cfg.dataset_manifest} is empty. Exiting!")
+            return None
+
         with open(cfg.dataset_manifest, 'r') as f:
+            partial_audio = True
             for line in f:
                 item = json.loads(line)
+                if "duration" not in item:
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:243]: duration is not the key factor determining limit on audio - it is both duration + offset presence that determines that partial audio is chosen. 

Duration is filled in for nearly all manifests always.

File: examples/asr/transcribe_speech.py [hunk 4]
```diff
@@ -151,10 +232,22 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
+                    partial_audio = False
                 filepaths.append(item['audio_filepath'])
+        if not partial_audio:
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:246]: Do not raise this warning - by default most manifest do not have offset set. We don't want to warm for a general case

File: examples/asr/transcribe_speech.py (hunk 2 above)
Comment (titu1994) [inline examples/asr/transcribe_speech.py:97]: These functions are quite large, better to move the into some util dir and import only here.

File: nemo/collections/asr/models/ctc_models.py [hunk 5]
```diff
@@ -691,9 +691,15 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
         Returns:
             A pytorch DataLoader for the given audio file(s).
         """
-        batch_size = min(config['batch_size'], len(config['paths2audio_files']))
+        if 'manifest_filepath' in config:
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:694]: What about RNNT overrides ? Add the support for it's two methods, even if it's not used yet. All transcribe related info should ideally be synchronized among all model implementations.

**Review by titu1994 — CHANGES_REQUESTED** [review]

**Comment by titu1994** [comment]
The notebook needs some grammar fixes + @tango4j  for review of the VAD code (it looks correct to me but second checks). 

Also, the method does not work for RNNT - the notebook needs to be renamed to ASR with VAD for CTC models (or some such equivalent title) + mention in the notebook that this is only for CTC at the moment. 
Also add this notebook link to the ASR tutorials readme https://github.com/NVIDIA/NeMo/tree/main/tutorials/asr under it's own section (ASR with VAD).

File: examples/asr/transcribe_speech.py (hunk 2 above)
Comment (fayejf) [reply examples/asr/transcribe_speech.py:97]: Cannot find a good existing place. How about having a transcribe_utils.py so we can put everything about transcribe there.

File: nemo/collections/asr/models/ctc_models.py (hunk 5 above)
Comment (fayejf) [reply nemo/collections/asr/models/ctc_models.py:694]: good point. I totally forgot rnnt. :(

**Comment by fayejf** [comment]
> Also, the method does not work for RNNT - the notebook needs to be renamed to ASR with VAD for CTC models (or some such equivalent title) + mention in the notebook that this is only for CTC at the moment. Also add this notebook link to the ASR tutorials readme https://github.com/NVIDIA/NeMo/tree/main/tutorials/asr under it's own section (ASR with VAD).

I added it to readme. [11] Is there any particular order for notebooks?
Oh, good point. it's should be Offline_ASR_with_VAD_for_CTC_models for now.

**Review by tango4j — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/utils/vad_utils.py [hunk 6]
```diff
@@ -881,3 +881,121 @@ def init_vad_model(model_path):
...
+) -> str:
+    """
+    Stitch the prediction of speech segments.
+    """
+
+    if not stitched_output_manifest:
+        stitched_output_manifest = "asr_stitched_output_manifest.json"
+
+    if not speech_segments_tensor_dir:
+        speech_segments_tensor_dir = "speech_segments"
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/parts/utils/vad_utils.py:897]: good point!

File: nemo/collections/asr/models/ctc_models.py (hunk 5 above)
Comment (fayejf) [reply nemo/collections/asr/models/ctc_models.py:694]: added support for rnnt and rnnt bpe

**Comment by fayejf** [comment]
@titu1994 @tango4j  Thanks for the careful review and feedback! I've addressed all the feedback above and did a grammar check again. Please review again when you have time. :)

File: docs/source/starthere/tutorials.rst [hunk 7]
```diff
@@ -88,6 +88,9 @@ To run a tutorial:
    * - ASR
      - Buffered Transducer inference with LCS Merge Algorithm
      - `Buffered Transducer inference with LCS Merge <https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/Buffered_Transducer_Inference_with_LCS_Merge.ipynb>`_
+   * - ASR
+     - Offline ASR with VAD for CTC models
+     - `Offline ASR with VAD for CTC models <https://github.com/NVIDIA/NeMo/blob/stable/tutorials/asr/Offline_ASR_with_VAD_for_CTC_models.ipynb>`_
```
Comment (titu1994) [inline docs/source/starthere/tutorials.rst:93]: Shouldn't this be a colab link ?

File: examples/asr/transcribe_speech.py [hunk 8]
```diff
@@ -151,10 +152,18 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
         filepaths = []
+        if os.stat(cfg.dataset_manifest).st_size == 0:
+            logging.error(f"The input dataset_manifest {cfg.dataset_manifest} is empty. Exiting!")
+            return None
+
         with open(cfg.dataset_manifest, 'r') as f:
+            partial_audio = True
             for line in f:
                 item = json.loads(line)
+                if "offset" not in item or "duration" not in item:
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:163]: This should be an if x in manifesf and y in manifest: situation, not an or. Most manifests will have duration but not offset.

**Review by tango4j — APPROVED** [review]

File: nemo/collections/asr/models/ctc_models.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:694]: Cool.

File: nemo/collections/asr/parts/utils/transcribe_utils.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    asr_model,
+    path2manifest: str,
+    batch_size: int = 4,
+    logprobs: bool = False,
+    return_hypotheses: bool = False,
+    num_workers: int = 0,
+) -> List[str]:
+
+    assert (
+        type(asr_model) == EncDecCTCModelBPE or type(asr_model) == EncDecCTCModel
```
Comment (titu1994) [inline nemo/collections/asr/parts/utils/transcribe_utils.py:33]: You can do is instance(asr_model, EncDecCTCModel) - subclassing should work out of the box.

File: examples/asr/transcribe_speech.py [hunk 10]
```diff
@@ -187,15 +196,25 @@ def autocast():
...
         with torch.no_grad():
-            transcriptions = asr_model.transcribe(filepaths, batch_size=cfg.batch_size, num_workers=cfg.num_workers)
+            if partial_audio:
+                transcriptions = transcribe_partial_audio(
+                    asr_model=asr_model,
+                    path2manifest=cfg.dataset_manifest,
+                    batch_size=cfg.batch_size,
+                    num_workers=cfg.num_workers,
+                )
+            else:
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:206]: Hmm we don't want to disallow users who have offset and duration in manifest from running transcribe (at least whole transcribe can work).  

Do one thing. Check for CTC here then if partial audio send to transcribe partial. If not CTC and still partial audio flag is set - log a warning message saying that RNNT models do not support partial audio transcription (for now) and instead perform the else part.

File: tutorials/asr/README.md [hunk 11]
```diff
@@ -28,6 +28,8 @@ In this repository, you will find several tutorials discussing what is Automatic
 
 10) `Self_Supervised_Pre_Training`: It can often be difficult to obtain labeled data for ASR training. In this tutorial, we demonstrate how to pre-train a speech model in an unsupervised manner, and then fine-tune with CTC loss.
 
+11) `Offline_ASR_with_VAD_for_CTC_models`: In this tutorial, we will demonstrate how to use offline VAD to extract speech segments and transcribe the speech segments with CTC models. This will help to exclude some non_speech utterances and could save computation resources by removing unnecessary input to the ASR system.
```
Comment (titu1994) [inline tutorials/asr/README.md:31]: Create another section called ASR with VAD instead of adding to current list. Then the table of content can quickly jump to it. We will add future RNNT too so might as well.

**Review by titu1994 — COMMENTED** [review]
Looks better. Minor modifications left.

File: tutorials/asr/README.md (hunk 11 above)
Comment (fayejf) [reply tutorials/asr/README.md:31]: Actually, I don't think we need to add another section.  RNNT will go to this tutorial by removing CTC models. And streaming vad asr will go to 

> Streaming / Buffered Automatic Speech Recognition: Streaming_ASR_with_VAD_for_CTC_models

File: docs/source/starthere/tutorials.rst (hunk 7 above)
Comment (fayejf) [reply docs/source/starthere/tutorials.rst:93]: good catch!

File: examples/asr/transcribe_speech.py (hunk 8 above)
Comment (fayejf) [reply examples/asr/transcribe_speech.py:163]: Not really. if anyone is missing, then partial_audio=False and we will use `audio_filepath` and `transcribe` method

**Comment by fayejf** [comment]
@titu1994 Replied and updated with review comments. Also, make necessary changes for smooth running on colab. Note the new link might be broken due to the 'stable' issue. 
We probably need to revert stable to main for main branch after each release.

File: examples/asr/transcribe_speech.py (hunk 8 above)
Comment (titu1994) [reply examples/asr/transcribe_speech.py:163]: We generally have the 3 values in the manifest - audio_filepath. text and duration. Here the or for "offset" will fail and set partial_audio = False which is fine.
The logic should be - partial audio is false by default, but if both args are provided, then set to true - its the inverse of this.

**Review by titu1994 — DISMISSED** [review]
Overall looks good. @tango4j  for final review before merge.

**Review by tango4j — APPROVED** [review]
